### PR TITLE
First code for the "events library"

### PR DESCRIPTION
**Pull request by jskeet**
This will be moved into its own repository eventually, but it's useful to develop it alongside the Functions Framework for right now.

These first commits introduce:

- The serialization-neutral library
- The Protobuf converter and types
- An extension method for Google.Protobuf.WellKnownTypes.Struct to make it (much) easier to work with

### CODE REVIEW

File: protos/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- product (e.g. `firebase` or `cloud.storage`)
+- resource (e.g. `object` or `document`)
+- version (e.g. `v1`)
+- event description (e.g. `finalized`, `deleted`)
+
+The proto package (in the `package` part of each .proto file) must
+be `google.events.`*product*`.`*version*`, and the .proto file
+should be in a corresponding directory. As a complete example,
+consider the event that is created when a Google Cloud Storage
+storage object is finalized:
```
[the remark this answers is not in the document]
Comment (jskeet) [reply protos/README.md]: Nope, this was entirely accidental :) However, the protos and external_protos directories weren't meant to be in this PR at all :( They're in a separate one for a different repo - I'll fix it there.

File: events-src/Google.Events.sln [hunk 2]
```diff
@@ -0,0 +1,51 @@
+﻿
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events.sln:1]: Nope, it's all in events-src. That will be `src` in the new repo.

File: events-src/Google.Events/CloudEventConverters.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            var converter = GetDataConverter<T>();
+            return converter.ConvertEventData(cloudEvent);
+        }
+
+
+        /// <summary>
+        /// Populates <paramref name="cloudEvent"/> with the data in <paramref name="data"/> by
+        /// using the <see cref="ICloudEventDataConverter{T}"/> associated with <typeparamref name="T"/> via
+        /// <see cref="CloudEventDataConverterAttribute"/>. See
+        /// <see cref="ICloudEventDataConverter{T}.PopulateCloudEvent(CloudEvent, T)"/> for details what gets populated.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events/CloudEventConverters.cs]: Done.

File: events-src/Google.Events/CloudEventConverters.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        /// </summary>
+        /// <param name="cloudEvent"></param>
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events/CloudEventConverters.cs]: Doh! Thank you :)

File: events-src/Google.Events/CloudEventConverters.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        /// <param name="data"></param>
+        public static void PopulateCloudEvent<T>(CloudEvent cloudEvent, T data) where T : class
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events/CloudEventConverters.cs:93]: I don't think so - let's trust the converter to do the right thing.

File: events-src/Google.Events/ICloudEventDataConverter.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    /// Implementations are expected to be thread-safe and stateless.
+    /// This is typically the type decorated with the <see cref="CloudEventDataConverterAttribute"/> attribute;
+    /// see the attribute documentation for additional expectations when the converter is discovered via the
+    /// attribute.
+    /// </summary>
+    /// <typeparam name="T">The type to deserialize CloudEvent data to.</typeparam>
+    public interface ICloudEventDataConverter<T> where T : class
+    {
+        /// <summary>
+        /// Converts the data in <paramref name="cloudEvent"/> to the type specified at construction
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events/ICloudEventDataConverter.cs]: Done.

File: events-src/Google.Events/ICloudEventDataConverter.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        /// the data content type as well. It should not be mutated by this method.</param>
+        /// <returns>The data from the CloudEvent, deserialized to <typeparamref name="T"/>.</returns>
+        T ConvertEventData(CloudEvent cloudEvent);
+
+        /// <summary>
+        /// Populates <paramref name="cloudEvent"/> with the data in <paramref name="data"/>.
+        /// </summary>
+        /// <remarks>
+        /// If <see cref="CloudEvent.DataContentType"/> is set within <paramref name="cloudEvent"/>, this
+        /// must not be changed, and should reflect the content type to use within the event. If no content
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events/ICloudEventDataConverter.cs:43]: No, it's broader than the method itself (although it *is* used in the method). It should affect what ends up in the `Data` property.

File: events-src/Google.Events.Protobuf/ProtobufCloudEventDataConverter.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        private static readonly ContentType s_protobufContentType = new ContentType("application/protobuf");
+        private static readonly JsonParser s_jsonParser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true));
+        private readonly MessageParser<T> _parser;
+        
+        /// <summary>
+        /// Constructs an event data converter for the Protobuf message <typeparamref name="T"/>.
+        /// </summary>
+        public ProtobufCloudEventDataConverter()
+        {
+            var parserProperty = typeof(T).GetProperty("Parser", BindingFlags.Static | BindingFlags.Public);
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events.Protobuf/ProtobufCloudEventDataConverter.cs]: Nope, unfortunately not. It's a static property. However, we *can* create an instance of T and then call MergeFrom. Hooray!
(Unfortunately that stops it from being a single expression, but never mind.)

File: events-src/Google.Events.Protobuf/StructExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        public static object GetValue(this Struct @struct, string path)
+        {
+            ProtoPreconditions.CheckNotNull(@struct, nameof(@struct));
+            ProtoPreconditions.CheckNotNull(path, nameof(path));
+            string[] segments = path.Split(s_slashSplit, StringSplitOptions.None);
+            Struct currentStruct = @struct;
+            Value currentValue = null; // We always go through the loop at least once, so this is okay.
+            foreach (var segment in segments)
+            {
+                if (currentStruct is null)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events.Protobuf/StructExtensions.cs:49]: I'd rather not do that yet - it feels a bit C#-specific, and I'm hoping we can get code for this everywhere.
But we can definitely introduce it later.

If the path so far has a value of null, I think it's fine to just throw when we try to go further.

File: events-src/Google.Events.Protobuf/StructExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                    throw new ArgumentException($"Path segment '{segment}' in path occurred when not in a struct value");
+                }
+                int indexStart = segment.IndexOf('[');
+                int indexEnd = segment.IndexOf(']');
+                // List handling
+                if (indexStart != -1 || indexEnd != -1)
+                {
+                    // There must be exactly one [, exactly one ], and
+                    // the [ must come before ].
+                    if (indexStart == -1 || indexEnd == -1 ||
```
[the remark this answers is not in the document]
Comment (jskeet) [reply events-src/Google.Events.Protobuf/StructExtensions.cs:60]: Ooh, that's a very good point. I'd expect it to be `bar[3][1]`. I'm going to add a TODO for the moment - I don't think it's high priority, and it's probably going to be fiddly, but I definitely want to come back to it. Nice catch!

File: events-src/Google.Events/CloudEventConverters.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply events-src/Google.Events/CloudEventConverters.cs:93]: OK

File: events-src/Google.Events.Protobuf/StructExtensions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply events-src/Google.Events.Protobuf/StructExtensions.cs:49]: I still think that it's going to be hard to check for nulls in the middle of the path (without catchting the exception that is, but that doesn't feel very C#). If I want to get to foo/bar/baz if it's there, then I get a null if baz is null but an exception if bar is null, so, to do that cleanly, I'd had to first get foo/bar, check for null and then foo/bar/baz.
But I'm fine with not doing that right now, this is still way better than before.

**Pull request merged**